test/compilable/test20868.d

```diff
@@ -0,0 +1,8 @@
+// https://issues.dlang.org/show_bug.cgi?id=20868
+// REQUIRED_ARGS: -dip1000
+
+void scoped (scope void delegate() dg)
+{
+    static void delegate()[] dgs;
+    dgs ~= dg;  // error
```

Comment:
Shouldn't `scope` apply to the context pointer?

`scope` checks are also not restricted to `@safe`, see e.g. 

```d
int* scoped(scope int* p) {
    return p;
}

void main() {
    int i;
    int* p = &i;
    scoped(p);
}
```
Compiling with `-dip1000`
```
onlineapp.d(2): Error: scope variable `p` may not be returned
```

Reply:
I'm not sure. Spec [1] says: "scope escape analysis is only done for @safe functions. For other functions scope semantics must be manually enforced.". Dip1000 [2] says: "Errors for scope violations are only reported in @safe code.".

However, there are cases like the one that you have highlighted where scope is enforced even if not in a @safe function.

[1] https://dlang.org/spec/function.html#scope-parameters
[2] https://github.com/dlang/DIPs/blob/master/DIPs/other/DIP1000.md

Reply:
Guess i dind't remember [1] (and [2] is outdated anyways).
But storage classes depending on function attributes doesn't seem right.

Reply:
@thewilsonator I am not sure if your review is right. Check the above comments. That is why we need some clarifications from @WalterBright .

Reply:
This test case being incorrect (i.e. compilable) has nothing to do with dip1000 and has everything to do with the original purpose of `scope delegate`s:

`scope` on a delegate means "this argument's context pointer does not outlive a call to this function, do not GC allocate a closure for it"

Reply:
Just to clarify, here is the original bug report:

```d
void scoped (scope void delegate() dg)
{
    static void delegate()[] dgs;
    dgs ~= dg;  // error

    nonScoped(dg);  // not an error???
}

void nonScoped (void delegate() dg)
{
    static void delegate()[] dgs;
    dgs ~= dg;
}

void main ()
{
    int x;
    scoped({x = 100;});
}
```

Without this patch:

1. If you compile without @safe or dip1000, you get no error.
2. If you compile with dip1000, you get an error at the line that has the error comment
3. If you compile with dip1000 and annotate the functions with @safe you get errors on both the line that has "error" comment and the line that has "non an error???" comment.

So, what is the correct behavior here? As far as the spec goes, scope should do nothing without -dip1000 and @safe, however, the language implementation sometimes does and sometimes doesn't respect that. So I am confused on what the fix should be or if there should be a fix at all.

If we consider the spec as the universal truth, then we should patch the compiler to ignore scope whenever we are not in a @safe function (non-dip1000 context).

@thewilsonator What is your source of truth?